File: frontend/src/api/endpoints/pos.ts
import { api } from "@/api/client";

// ─────────────────────────────────────────────────────────────────────────────
// OFBiz Point-of-Sale (POS) — endpoint wrappers + inline TS types.
// Mirrors app/routers/pos.py (LIVE) + app/models.py POS models.
// All amounts are integer cents.
// ─────────────────────────────────────────────────────────────────────────────

export interface RegisterConfig {
  register_id?: string | null;
  label: string;
  location_id: string;
  default_currency: string;
  created_at: number;
  updated_at: number;
  created_by?: string | null;
}

export interface RegisterCreateIn {
  label: string;
  location_id: string;
  default_currency?: string;
}

export type TenderKind = "cash" | "card" | "wallet";

export interface TenderOut {
  kind: string; // cash | card | wallet
  amount_cents: number;
  change_due_cents: number; // only meaningful for kind=cash
  payment_method_id?: string | null;
  card_ref?: string | null;
}

export interface RegisterSessionOut {
  session_id: string;
  register_id: string;
  cashier_sub: string;
  status: string; // open | closed
  opening_float_cents: number;
  closing_float_cents?: number | null;
  expected_cash_cents?: number | null;
  counted_cash_cents?: number | null;
  over_short_cents?: number | null; // signed: positive = over
  opened_at: number;
  closed_at?: number | null;
}

export interface PosTransactionOut {
  txn_id: string;
  session_id: string;
  cart_id?: string | null;
  order_id?: string | null;
  status: string; // draft | tendered | voided | refunded
  subtotal_cents: number;
  discount_cents: number;
  tax_cents: number;
  total_cents: number;
  tenders: TenderOut[];
  receipt_id?: string | null;
  created_at: number;
  updated_at: number;
}

export interface PosTxnDraftOut {
  txn_id: string;
  session_id: string;
  cashier_sub: string;
  status: string;
  cart_id: string;
  subtotal_cents: number;
  tax_cents: number;
  discount_cents: number;
  total_cents: number;
  created_at: number;
  updated_at: number;
}

/** Cart line item shape (from shoppingcart._item_from_item). */
export interface PosLineItem {
  sku: string;
  name: string;
  quantity: number;
  unit_price_cents: number;
  line_total_cents: number;
  updated_at?: number | null;
  image_url?: string;
  category_id?: string;
  item_id?: string;
}

/** Response from add/set/remove line item helpers. */
export interface PosLineMutationOut {
  txn_id: string;
  total_cents: number;
  item?: PosLineItem;
}

export interface OpenSessionIn {
  register_id: string;
  opening_float_cents?: number;
  idempotency_key?: string | null;
}

export interface CloseSessionIn {
  counted_cash_cents: number;
}

export interface PosBindTxnIn {
  correlation_id: string;
}

export interface PosAddLineItemIn {
  item_id?: string | null;
  category_id?: string | null;
  sku?: string | null;
  name?: string | null;
  unit_price_cents?: number | null;
  quantity?: number;
}

export interface PosSetLineQtyIn {
  quantity: number;
}

export interface PosCashTenderIn {
  amount_tendered_cents: number;
  idempotency_key: string;
}

export interface PosCardTenderIn {
  amount_tendered_cents: number;
  payment_method_id: string;
  idempotency_key: string;
  card_kind?: string | null;
  last4?: string | null;
}

export interface PosWalletTenderIn {
  amount_tendered_cents: number;
  idempotency_key: string;
}

export interface PosTxnVoidIn {
  reason: string;
}

export interface RefundTxnIn {
  txn_id: string;
  reason: string;
}

export interface ReceiptLinkOut {
  txn_id: string;
  receipt_path: string;
  receipt_url: string;
  generated_at: number;
}

export interface PosSessionReportOut {
  session_id: string;
  register_id: string;
  cashier_sub: string;
  status: string;
  opening_float_cents: number;
  closing_float_cents?: number | null;
  expected_cash_cents?: number | null;
  counted_cash_cents?: number | null;
  over_short_cents?: number | null;
  opened_at: number;
  closed_at?: number | null;
  transaction_count: number;
  total_sales_cents: number;
  total_cash_tendered_cents: number;
  total_change_given_cents: number;
}

// ── X/Z till-summary report models (POS-010) ──────────────────────────────────

export interface TenderBreakdownItem {
  kind: string; // cash | card | wallet
  gross_cents: number;
  refund_cents: number;
  net_cents: number;
  transaction_count: number;
}

export interface SessionReportCashSection {
  opening_float_cents: number;
  cash_in_cents: number;
  change_out_cents: number;
  expected_cash_cents: number;
  counted_cash_cents?: number | null;
  over_short_cents?: number | null;
}

export interface SessionReportOut {
  report_kind: string; // "x" | "z"
  session_id: string;
  register_id: string;
  cashier_sub: string;
  opened_at: number;
  closed_at?: number | null;
  generated_at: number;
  z_report_printed_at?: number | null;
  gross_sales_cents: number;
  discount_cents: number;
  tax_cents: number;
  net_sales_cents: number;
  refund_total_cents: number;
  void_count: number;
  transaction_count: number;
  tender_breakdown: TenderBreakdownItem[];
  cash: SessionReportCashSection;
}

// ─── Registers ────────────────────────────────────────────────────────────────

export const createRegister = (body: RegisterCreateIn) =>
  api.post<RegisterConfig>("/ui/pos/registers", body);

export const getRegister = (registerId: string) =>
  api.get<RegisterConfig>(`/ui/pos/registers/${registerId}`);

// ─── Sessions ───────────────────────────────────────────────────────────────

export const openSession = (body: OpenSessionIn) =>
  api.post<RegisterSessionOut>("/ui/pos/sessions", body);

export const getOpenSession = (registerId: string) =>
  api.get<RegisterSessionOut>(`/ui/pos/registers/${registerId}/open-session`);

export const getSession = (sessionId: string) =>
  api.get<RegisterSessionOut>(`/ui/pos/sessions/${sessionId}`);

export const closeSession = (sessionId: string, body: CloseSessionIn) =>
  api.post<RegisterSessionOut>(`/ui/pos/sessions/${sessionId}/close`, body);

// ─── Session reports ──────────────────────────────────────────────────────────

export const getSessionReport = (sessionId: string) =>
  api.get<PosSessionReportOut>(`/ui/pos/sessions/${sessionId}/report`);

export const getXReport = (sessionId: string) =>
  api.get<SessionReportOut>(`/ui/pos/sessions/${sessionId}/report/x`);

export const getZReport = (sessionId: string) =>
  api.get<SessionReportOut>(`/ui/pos/sessions/${sessionId}/report/z`);

export const listSessionTxns = (sessionId: string) =>
  api.get<PosTransactionOut[]>(`/ui/pos/sessions/${sessionId}/transactions`);

// ─── Transaction drafts (cart binding + line items) ───────────────────────────

export const bindTxn = (sessionId: string, body: PosBindTxnIn) =>
  api.post<PosTxnDraftOut>(`/ui/pos/sessions/${sessionId}/txns`, body);

export const getTxn = (txnId: string) =>
  api.get<PosTxnDraftOut>(`/ui/pos/txns/${txnId}`);

export const addLineItem = (txnId: string, body: PosAddLineItemIn) =>
  api.post<PosLineMutationOut>(`/ui/pos/txns/${txnId}/lines`, body);

export const listTxnLines = (txnId: string) =>
  api.get<PosLineItem[]>(`/ui/pos/txns/${txnId}/lines`);

export const setLineQty = (txnId: string, sku: string, body: PosSetLineQtyIn) =>
  api.patch<PosLineMutationOut>(`/ui/pos/txns/${txnId}/lines/${encodeURIComponent(sku)}`, body);

export const removeLineItem = (txnId: string, sku: string, decrement = 1) =>
  api.del<PosLineMutationOut>(`/ui/pos/txns/${txnId}/lines/${encodeURIComponent(sku)}`, {
    decrement: String(decrement),
  });

// ─── Tender / void / refund ───────────────────────────────────────────────────

export const tenderCash = (sessionId: string, txnId: string, body: PosCashTenderIn) =>
  api.post<PosTransactionOut>(
    `/ui/pos/sessions/${sessionId}/transactions/${txnId}/tender-cash`,
    body,
  );

export const tenderCard = (sessionId: string, txnId: string, body: PosCardTenderIn) =>
  api.post<PosTransactionOut>(
    `/ui/pos/sessions/${sessionId}/transactions/${txnId}/tender-card`,
    body,
  );

export const tenderWallet = (sessionId: string, txnId: string, body: PosWalletTenderIn) =>
  api.post<PosTransactionOut>(
    `/ui/pos/sessions/${sessionId}/transactions/${txnId}/tender-wallet`,
    body,
  );

export const voidTxn = (txnId: string, body: PosTxnVoidIn) =>
  api.post<PosTransactionOut>(`/ui/pos/txns/${txnId}/void`, body);

export const refundTxn = (txnId: string, body: RefundTxnIn) =>
  api.post<PosTransactionOut>(`/ui/pos/txns/${txnId}/refund`, body);

// ─── Receipt ──────────────────────────────────────────────────────────────────

export const getReceiptLink = (txnId: string) =>
  api.get<ReceiptLinkOut>(`/ui/pos/txns/${txnId}/receipt`);

/** Absolute path for the receipt PDF (download / preview). */
export const receiptPdfUrl = (txnId: string) =>
  `/ui/pos/txns/${txnId}/receipt/pdf`;

// ─── Currency helper ──────────────────────────────────────────────────────────

export function fmtCents(cents: number | null | undefined, currency = "USD"): string {
  const v = (cents ?? 0) / 100;
  try {
    return new Intl.NumberFormat(undefined, { style: "currency", currency }).format(v);
  } catch {
    return `$${v.toFixed(2)}`;
  }
}

File: app/routers/pos.py
"""
POS router — /ui/pos/* endpoints.

Mounted unconditionally in app/main.py; every handler calls _require_enabled()
to return 404 when S.pos_enabled is False.  Follows the pattern established by
app/routers/inventory.py and app/routers/returns_rma.py.
"""
from __future__ import annotations

from typing import Any, Dict, List

from fastapi import APIRouter, Depends, HTTPException, Query, Response

from app.services.sessions import require_ui_session
from app.auth.policy import require_admin_or_root_csrf
from app.services.pos_register import (
    _require_enabled,
    create_register,
    get_register,
    open_session,
    get_open_session,
    get_session,
    close_session,
    bind_txn_draft,
    get_txn_draft,
    add_line_item,
    set_line_item_qty,
    remove_line_item,
    list_txn_items,
)
from app.services.pos_tender import (
    tender_cash,
    tender_card,
    tender_wallet,
    void_txn,
    refund_txn,
    get_session_report,
    list_session_txns,
)
from app.services.pos_reports import session_report
from app.services.pos_receipt import (
    get_or_create_pos_receipt,
    get_pos_receipt_bytes,
)
from app.models import (
    RegisterConfig,
    RegisterCreateIn,
    RegisterSessionOut,
    OpenSessionIn,
    CloseSessionIn,
    PosTransactionOut,
    PosTxnDraftOut,
    PosBindTxnIn,
    PosAddLineItemIn,
    PosSetLineQtyIn,
    PosCashTenderIn,
    PosCardTenderIn,
    PosWalletTenderIn,
    PosTxnVoidIn,
    RefundTxnIn,
    PosSessionReportOut,
    ReceiptLinkOut,
    SessionReportOut,
)

pos_router = APIRouter(prefix="/ui/pos", tags=["pos"])


# ── registers ─────────────────────────────────────────────────────────────────

@pos_router.post("/registers", response_model=RegisterConfig)
async def create_register_endpoint(
    body: RegisterCreateIn,
    session: Dict[str, Any] = Depends(require_admin_or_root_csrf),
) -> Dict[str, Any]:
    _require_enabled()
    return create_register(
        user_sub=session["user_sub"],
        label=body.label,
        location_id=body.location_id,
        default_currency=body.default_currency,
    )


@pos_router.get("/registers/{register_id}", response_model=RegisterConfig)
async def get_register_endpoint(
    register_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    reg = get_register(register_id)
    if not reg:
        raise HTTPException(status_code=404, detail="Register not found")
    return reg


# ── sessions ──────────────────────────────────────────────────────────────────

@pos_router.post("/sessions", response_model=RegisterSessionOut)
async def open_session_endpoint(
    body: OpenSessionIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return open_session(
        cashier_sub=session["user_sub"],
        register_id=body.register_id,
        opening_float_cents=body.opening_float_cents,
        idempotency_key=body.idempotency_key,
    )


@pos_router.get("/registers/{register_id}/open-session", response_model=RegisterSessionOut)
async def get_open_session_endpoint(
    register_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    result = get_open_session(register_id)
    if not result:
        raise HTTPException(status_code=404, detail="No open session for this register")
    return result


@pos_router.get("/sessions/{session_id}", response_model=RegisterSessionOut)
async def get_session_endpoint(
    session_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    result = get_session(session_id)
    if not result:
        raise HTTPException(status_code=404, detail="Session not found")
    return result


@pos_router.post("/sessions/{session_id}/close", response_model=RegisterSessionOut)
async def close_session_endpoint(
    session_id: str,
    body: CloseSessionIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    # Ownership check: cashier may only close their own session unless admin/root.
    sess = get_session(session_id)
    if sess is None:
        raise HTTPException(status_code=404, detail="Session not found")

    from app.auth.roles import Role
    user_role = session.get("role", Role.USER)
    is_admin = hasattr(user_role, "value") and user_role.value >= Role.ADMIN.value
    if not is_admin and sess.get("cashier_sub") != session["user_sub"]:
        raise HTTPException(status_code=403, detail="You may only close your own session")

    return close_session(
        session_id=session_id,
        counted_cash_cents=body.counted_cash_cents,
        actor_sub=session["user_sub"],
    )


# ── session reports ───────────────────────────────────────────────────────────

@pos_router.get("/sessions/{session_id}/report", response_model=PosSessionReportOut)
async def session_report_endpoint(
    session_id: str,
    session: Dict[str, Any] = Depends(require_admin_or_root_csrf),
) -> Dict[str, Any]:
    _require_enabled()
    return get_session_report(session_id)


@pos_router.get("/sessions/{session_id}/report/x", response_model=SessionReportOut)
async def session_x_report_endpoint(
    session_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return session_report(
        session_id,
        "x",
        cashier_sub=session["user_sub"],
        role=session.get("role", "user"),
    )


@pos_router.get("/sessions/{session_id}/report/z", response_model=SessionReportOut)
async def session_z_report_endpoint(
    session_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return session_report(
        session_id,
        "z",
        cashier_sub=session["user_sub"],
        role=session.get("role", "user"),
    )


@pos_router.get("/sessions/{session_id}/transactions", response_model=List[PosTransactionOut])
async def list_session_txns_endpoint(
    session_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> List[Dict[str, Any]]:
    _require_enabled()
    return list_session_txns(session_id)


# ── transaction drafts (cart binding + line items) ────────────────────────────

@pos_router.post("/sessions/{session_id}/txns", response_model=PosTxnDraftOut)
async def bind_txn_endpoint(
    session_id: str,
    body: PosBindTxnIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return bind_txn_draft(
        cashier_sub=session["user_sub"],
        session_id=session_id,
        correlation_id=body.correlation_id,
    )


@pos_router.get("/txns/{txn_id}", response_model=PosTxnDraftOut)
async def get_txn_endpoint(
    txn_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return get_txn_draft(cashier_sub=session["user_sub"], txn_id=txn_id)


# ── receipt (POS-009) ─────────────────────────────────────────────────────────

def _require_txn_owner_or_admin(txn_id: str, session: Dict[str, Any]) -> None:
    """Cashier may only access their own txn's receipt; admin/root any."""
    from app.core.tables import T
    from app.auth.roles import Role

    item = T.pos.get_item(Key={"pos_pk": f"TXN#{txn_id}", "pos_sk": "META"}).get("Item")
    if not item:
        raise HTTPException(status_code=404, detail="Transaction not found")
    user_role = session.get("role", Role.USER)
    role_val = getattr(user_role, "value", user_role)
    is_admin = role_val in ("admin", "root")
    if not is_admin and item.get("cashier_sub") != session["user_sub"]:
        raise HTTPException(status_code=403, detail="You may only access your own receipts")


@pos_router.get("/txns/{txn_id}/receipt", response_model=ReceiptLinkOut)
async def get_txn_receipt_endpoint(
    txn_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    _require_txn_owner_or_admin(txn_id, session)
    return get_or_create_pos_receipt(txn_id)


@pos_router.get("/txns/{txn_id}/receipt/pdf")
async def get_txn_receipt_pdf_endpoint(
    txn_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Response:
    _require_enabled()
    _require_txn_owner_or_admin(txn_id, session)
    pdf_bytes = get_pos_receipt_bytes(txn_id)
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'attachment; filename="pos-receipt-{txn_id}.pdf"',
        },
    )


@pos_router.post("/txns/{txn_id}/lines")
async def add_line_item_endpoint(
    txn_id: str,
    body: PosAddLineItemIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return add_line_item(
        cashier_sub=session["user_sub"],
        txn_id=txn_id,
        sku=body.sku,
        name=body.name,
        unit_price_cents=body.unit_price_cents,
        item_id=body.item_id,
        category_id=body.category_id,
        quantity=body.quantity,
    )


@pos_router.get("/txns/{txn_id}/lines")
async def list_txn_lines_endpoint(
    txn_id: str,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> List[Dict[str, Any]]:
    _require_enabled()
    return list_txn_items(cashier_sub=session["user_sub"], txn_id=txn_id)


@pos_router.patch("/txns/{txn_id}/lines/{sku}")
async def set_line_qty_endpoint(
    txn_id: str,
    sku: str,
    body: PosSetLineQtyIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return set_line_item_qty(
        cashier_sub=session["user_sub"],
        txn_id=txn_id,
        sku=sku,
        quantity=body.quantity,
    )


@pos_router.delete("/txns/{txn_id}/lines/{sku}")
async def remove_line_item_endpoint(
    txn_id: str,
    sku: str,
    decrement: int = 1,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return remove_line_item(
        cashier_sub=session["user_sub"],
        txn_id=txn_id,
        sku=sku,
        decrement=decrement,
    )


# ── tender / void / refund ────────────────────────────────────────────────────

@pos_router.post("/sessions/{session_id}/transactions/{txn_id}/tender-cash",
                 response_model=PosTransactionOut)
async def tender_cash_endpoint(
    session_id: str,
    txn_id: str,
    body: PosCashTenderIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return tender_cash(
        session_id=session_id,
        txn_id=txn_id,
        cashier_sub=session["user_sub"],
        amount_tendered_cents=body.amount_tendered_cents,
        idempotency_key=body.idempotency_key,
    )


@pos_router.post("/sessions/{session_id}/transactions/{txn_id}/tender-card",
                 response_model=PosTransactionOut)
async def tender_card_endpoint(
    session_id: str,
    txn_id: str,
    body: PosCardTenderIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return tender_card(
        session_id=session_id,
        txn_id=txn_id,
        cashier_sub=session["user_sub"],
        amount_tendered_cents=body.amount_tendered_cents,
        payment_method_id=body.payment_method_id,
        idempotency_key=body.idempotency_key,
        card_kind=body.card_kind,
        last4=body.last4,
    )


@pos_router.post("/sessions/{session_id}/transactions/{txn_id}/tender-wallet",
                 response_model=PosTransactionOut)
async def tender_wallet_endpoint(
    session_id: str,
    txn_id: str,
    body: PosWalletTenderIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return tender_wallet(
        session_id=session_id,
        txn_id=txn_id,
        cashier_sub=session["user_sub"],
        amount_tendered_cents=body.amount_tendered_cents,
        idempotency_key=body.idempotency_key,
    )


@pos_router.post("/txns/{txn_id}/void", response_model=PosTransactionOut)
async def void_txn_endpoint(
    txn_id: str,
    body: PosTxnVoidIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return void_txn(
        txn_id=txn_id,
        cashier_sub=session["user_sub"],
        reason=body.reason,
    )


@pos_router.post("/txns/{txn_id}/refund", response_model=PosTransactionOut)
async def refund_txn_endpoint(
    txn_id: str,
    body: RefundTxnIn,
    session: Dict[str, Any] = Depends(require_ui_session),
) -> Dict[str, Any]:
    _require_enabled()
    return refund_txn(
        txn_id=txn_id,
        cashier_sub=session["user_sub"],
        reason=body.reason,
    )
